app/Http/Controllers/Ajax/DeviceSearchController.php

```diff
@@ -0,0 +1,108 @@
+<?php
+/*
+ * Search.php
+ *
+ * -Description-
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @package    LibreNMS
+ * @link       http://librenms.org
+ * @copyright  2021 Tony Murray
+ * @author     Tony Murray <murraytony@gmail.com>
+ */
+
+namespace App\Http\Controllers\Ajax;
+
+use App\Models\Device;
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Http\Request;
+use LibreNMS\Config;
+use LibreNMS\Util\Color;
+
+class DeviceSearchController extends SearchController
+{
+    public function buildQuery(string $search, Request $request): Builder
+    {
+        $baseQuery = Device::hasAccess($request->user())
+            ->leftJoin('locations', 'locations.id', '=', 'devices.location_id')
+            ->select(['devices.*', 'locations.location'])
+            ->distinct()
+            ->orderBy('devices.hostname');
+
+        return $baseQuery
+            ->where(function (Builder $query) use ($search, $baseQuery) {
+                // search filter
+                $index_friendly_like_search = "$search%";
+                $like_search = "%$search%";
+                $query->orWhere('hostname', 'LIKE', $index_friendly_like_search)
+                    ->orWhere('sysName', 'LIKE', $index_friendly_like_search)
+                    ->orWhere('display', 'LIKE', $index_friendly_like_search)
+                    ->orWhere('hostname', 'LIKE', $like_search)
+                    ->orWhere('sysName', 'LIKE', $like_search)
+                    ->orWhere('display', 'LIKE', $like_search)
+                    ->orWhere('location', 'LIKE', $like_search)
+                    ->orWhere('purpose', 'LIKE', $like_search)
+                    ->orWhere('serial', 'LIKE', $like_search)
+                    ->orWhere('notes', 'LIKE', $like_search);
+
+                if (\LibreNMS\Util\IPv4::isValid($search, false)) {
+                    $baseQuery->leftJoin('ports', 'ports.device_id', '=', 'devices.device_id')
+                        ->leftJoin('ipv4_addresses', 'ipv4_addresses.port_id', 'ports.port_id');
+
+                    $query->orWhere('ipv4_address', '=', $search)
+                        ->orWhere('overwrite_ip', '=', $search)
+                        ->orWhere('ip', '=', inet_pton($search));
```

Comment:
This is now exact matching instead of wildcard, but the `isValid()` check above prevents it anyway;
```
>>> \LibreNMS\Util\IPv4::isValid("10.10", false);
=> false
```
I think that will be annoying

Reply:
@Jellyfrog copied from the existing code.... The joins are painful.  Additionally, the ip field is binary, so it won't match partials without a lot of extra logic.

I was thinking of moving the interface IP stuff to the port search, but decided that would be a separate PR.

Reply:
Hmm

$query_filter .= ' OR `V4`.`ipv4_address` LIKE ?
                                        OR `D`.`overwrite_ip` LIKE ?
                                        OR `D`.`ip` = ? ';
                 $query_args_list = array_merge($query_args_list, ["%$search%", "%$search%", 

Well it's a mix since it's wrapped in a validip check. 

Think we should change to some regex that just checks if it's something that looks like a ip. Ie numbers and dots :)

Reply:
Sounds good, but I'd prefer a separate PR for that :)